### PR TITLE
Static snippets

### PR DESCRIPTION
**Pull request by rikasun**
https://user-images.githubusercontent.com/40652126/222879141-f530d531-cc19-4ae4-a91e-da8b38c3e5b7.mp4

### CODE REVIEW

File: libs/ui/src/lib/snippets/list-item/list-item.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      }),
+    [active],
+  );
+
+  return (
+    <button
+      onMouseOver={() => setActive(true)}
+      onMouseOut={() => setActive(false)}
+      onFocus={() => {}}
+      onBlur={() => {}}
```
Comment (ashblue) [inline libs/ui/src/lib/snippets/list-item/list-item.tsx:41]: Shouldn't need to override these events. Guessing leftover copilot code.

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    </div>
+  );
+}
+
+export default function Snippets() {
+  const [showSnippet, setShowSnippet] = useState(false);
+  const [q, setQ] = useState('');
+  const SnippetsCategoryList = useMemo(
+    () =>
+      ['GitHub', 'Kubernetes', 'Uncategorized'].map((name) => (
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/snippets.tsx:59]: Good data driven design here

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    <div className="fixed z-10">
+      <SnippetDialog showModal onOpenChange={() => {}}>
+        <div className="w-96 rounded-sm border border-zinc-700 bg-zinc-900 text-white">
+          <div className="flex">
+            <div className="w-60 border-r border-zinc-700 bg-zinc-800">
+              <div className="py-2">
+                <SnippetCategory
+                  name="All"
+                  selectedCategory="All"
+                  onClick={() => {}}
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/snippets.tsx:79]: Why is the onclick blank?

**Review by ashblue — CHANGES_REQUESTED** [review]
* Swap JS hovers in `list-item` for CSS classes
* Swap radix dialogue library for a simple modal pattern
  * Move modal logic into the app unless we're going to use that specific modal multiple times. Probably worth creating a re-usable modal at some point in the future. But probably too much to do right now. If we were to do a re-usable modal it should look something like this https://mui.com/material-ui/react-modal/

File: libs/ui/src/lib/snippets/list-item/list-item.tsx (hunk 1 above)
Comment (rikasun) [reply libs/ui/src/lib/snippets/list-item/list-item.tsx:41]: it was typescript but I don't need them anymore after using tailwind css only.

File: libs/ui/src/lib/snippets/list-item/list-item.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      onClick={onClick}
+      className={containerClasses}
+    >
```
[the remark this answers is not in the document]
Comment (rikasun) [reply libs/ui/src/lib/snippets/list-item/list-item.tsx]: Good catch! updated in https://github.com/cased/shell/pull/82/commits/3ce8495ecc4253a150479d22bef5963164bfd8a6

File: libs/ui/src/lib/snippets/dialog/dialog.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import * as Dialog from '@radix-ui/react-dialog';
```
[the remark this answers is not in the document]
Comment (rikasun) [reply libs/ui/src/lib/snippets/dialog/dialog.tsx]: Updated in https://github.com/cased/shell/pull/82/commits/9bbc3decbef278ef4d94bfa076400928509eb614. I also removed the library.

File: apps/shell/src/app/modules/prompt/snippets/snippets.scss [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  border-radius: 6px;
+}
+
+.snippets-content:focus {
+  outline: none;
+}
+
+.snippet-search {
+  height: 36px;
+}
```
[the remark this answers is not in the document]
Comment (rikasun) [reply apps/shell/src/app/modules/prompt/snippets/snippets.scss]: This is gone after we remove radix. The css `modal modal-open` seems to work!

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                  onClick={() => setShowSnippet(true)}
+                />
+              </div>
+            )}
+          </div>
+
+          <div className="w-full border-t border-zinc-700">
+            <form onSubmit={() => {}}>
+              <input type="submit" className="hidden" />
+              <input
```
[the remark this answers is not in the document]
Comment (rikasun) [reply apps/shell/src/app/modules/prompt/snippets/snippets.tsx]: Added in https://github.com/cased/shell/pull/82/commits/b8d0d9956780cb23975d466f2b1bcd6399956e60

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx (hunk 3 above)
Comment (rikasun) [reply apps/shell/src/app/modules/prompt/snippets/snippets.tsx]: I wasn't ready to put in local state or redux then but I added in https://github.com/cased/shell/pull/82/commits/e2e185bd3f02bede4250af3c9b7bf704eaabbc7d now

File: apps/shell/src/app/modules/prompt/snippets/category.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  className,
+  onClick,
+  selectedCategory,
+}: SnippetCategoryProps) {
+  const classNames = useMemo(
+    () =>
+      clsx(
+        'block py-0.5 px-3 focus:outline-none hover:bg-zinc-700 w-full text-left',
+        selectedCategory === name && 'font-bold',
+        className,
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/category.tsx:22]: I think you can use the object syntax for readability here.

```suggestion
        className,
        {
          'font-bold': selectedCategory === name,
        },
```

File: package.json [hunk 9]
```diff
@@ -54,7 +54,6 @@
     "@headlessui/react": "^1.6.1",
     "@heroicons/react": "^2.0.12",
     "@observablehq/plot": "^0.5.2",
-    "@radix-ui/react-dialog": "^0.1.1",
```
Comment (ashblue) [inline package.json:57]: Super duper thank you for removing this. Large dependencies lightly used like this are always a little scary for maintenance.

File: apps/shell/src/app/modules/prompt/snippets/list-item.tsx [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+export function SnippetListItem({
+  name,
+  description,
+  onClick,
+}: SnippetListItemProps) {
+  return (
+    <button
+      onClick={onClick}
+      className="w-full border-t border-zinc-800 py-2 px-4 text-left first-of-type:border-0 hover:cursor-pointer hover:bg-blue-700 hover:text-white"
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/list-item.tsx:15]: Thank you for moving all those class names into the component. I'm sure it wasn't easy converting all that old SCSS code.

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        <div>
+          <label htmlFor="label1" className="mb-1 mr-2 text-sm">
+            label1
+          </label>
+          <input
+            type="text"
+            className="border border-zinc-600 bg-transparent text-sm focus:bg-transparent focus:outline-none"
+            value="input1"
+            onChange={() => {}}
+          />
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/snippets.tsx:24]: There's actually a much easier way to structure labels (see below). Also try to think of the label name in two ways.

* Screen readers will hear this label name for anybody with a disability who is trying to find a specific input
* The name will help greatly with E2Es and testing. As the label can be used as a pseudo ID to find an input

That said, here are some proposed tweaks below that could help with usability on this. Not sure this is 100% correct, but worth considering.

```suggestion
          <label>
            {/* adding sr-only will hide the text safely, but still make it usable for tests and screen readers */}
            <span className="mb-1 mr-2 text-sm">Argument 1</span>
            <input
              type="text"
              className="border border-zinc-600 bg-transparent text-sm focus:bg-transparent focus:outline-none"
              value="input1"
              onChange={() => {}}
            />
          </label>
```

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        </pre>
+      </div>
+      <Button display="primary" size="small" onClick={() => {}}>
+        Run
+      </Button>
+    </div>
+  );
+}
+
+export default function Snippets() {
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/snippets.tsx:52]: Might be worth moving this to a separate file. File is getting a bit long (lots of HTML).

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+                />
+              </div>
+            )}
+          </div>
+
+          <div className="w-full border-t border-zinc-700">
+            <form onSubmit={() => {}}>
+              <label>
+                <input type="submit" className="hidden" />
+              </label>
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/snippets.tsx:110]: This syntax looks odd. Shouldn't this be a button type submit instead?

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+              <label>
+                <input
```
Comment (ashblue) [inline apps/shell/src/app/modules/prompt/snippets/snippets.tsx:112]:
```suggestion
              <label>
                <span className="sr-only">Search Snippets</span>
                <input
```

File: libs/ui/src/lib/modal/modal.tsx [hunk 15]
```diff
@@ -0,0 +1,16 @@
+import clsx from 'clsx';
+import { useMemo } from 'react';
+
+export interface IModal {
+  open: boolean;
+  children: React.ReactNode;
+}
+
+export function Modal({ open, children }: IModal) {
```
Comment (ashblue) [inline libs/ui/src/lib/modal/modal.tsx:9]: Nice! We can definitely use this as a component to build on. We can add things like `<ModalHeader />`, `<ModalClose />`, `<Box />`, etc as we go. And we can leave the styling to Ben when he wants to globally tweak all the modals.

File: libs/ui/src/lib/modal/modal.tsx [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+  const modalClass = useMemo(
+    () =>
+      clsx('modal bg-black/50', { 'modal-open': open, 'modal-closed': !open }),
```
Comment (ashblue) [inline libs/ui/src/lib/modal/modal.tsx:12]: Looks like we're using the Daisy UI modal syntax here. What about just using Tailwind instead so we have more granular control? Something like this maybe?

```suggestion
      clsx('fixed top-0 left-0 w-full h-full bg-black/50', { 'modal-open': open, 'modal-closed': !open }),
```

File: libs/ui/src/lib/modal/modal.tsx [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    [open],
+  );
+  return <div className={modalClass}>{children}</div>;
```
Comment (ashblue) [inline libs/ui/src/lib/modal/modal.tsx:15]: Might be worth adding a close click area here. That way clicking outside the content will close the window automatically.

```suggestion
  const [isOpen, setIsOpen] = useState(open);
  
  useEffect(() => {
    setIsOpen(open);
  }, [open]);

  return <div className={modalClass}>
    <div onClick={() => setIsOpen(false)} className="absolute top-0 left-0 w-full h-full" />
    {children}
  </div>;
```

Above suggestion should give you two way binding control. If that gives you any trouble with your current implementation it's okay to skip on this for now.

**Review by ashblue — APPROVED** [review]
Minor feedback and a handful of comments to hopefully spark some fun conversations on design architecture. Great job on this!!! :shipit:

File: libs/ui/src/lib/modal/modal.tsx (hunk 16 above)
Comment (rikasun) [reply libs/ui/src/lib/modal/modal.tsx]: I was wondering which lib these `css` are from in the permission modal!😅 If we remove`modal`, the `modal-open` and `modal-close` doesn't seem to work so I put modal back for now.
Do we want to use the css in https://tailwind-elements.com/docs/standard/components/modal/?

File: libs/ui/src/lib/modal/modal.tsx (hunk 17 above)
Comment (rikasun) [reply libs/ui/src/lib/modal/modal.tsx]: I was wondering how to do this. This works great! But we just need to coordinate the z-index we applied to different layers in the future, so far
xterm: z-10
close area: z-20
command pallet: z-50

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx (hunk 13 above)
Comment (rikasun) [reply apps/shell/src/app/modules/prompt/snippets/snippets.tsx:111]: Updated to button. Lots of signup/login form seem to use submit input.

File: libs/ui/src/lib/modal/modal.tsx (hunk 16 above)
Comment (ashblue) [reply libs/ui/src/lib/modal/modal.tsx]: I was wondering about that. All good, we can start with Daisy UI and transition off. We actually don't use the tailwind component library though. Sadly we don't have access to that CSS link.

File: libs/ui/src/lib/modal/modal.tsx (hunk 17 above)
Comment (ashblue) [reply libs/ui/src/lib/modal/modal.tsx]: Yeah the parent class will probably need a Z index. We should probably figure out some sorta strategy to our z index layers 🤔 Maybe Ben has a good idea on how to best organize this as Tailwind's z index layers are pretty vague.

File: apps/shell/src/app/modules/prompt/snippets/snippets.tsx (hunk 13 above)
Comment (ashblue) [reply apps/shell/src/app/modules/prompt/snippets/snippets.tsx]: You're right that a lot of forms use that pattern. But it's the old way of approaching submit buttons (pre HTML5). Using a modern button is better for a lot of reasons 😄

**Pull request merged**